### DISCUSSION TITLE
How to prevent fastify from logging authorization errors?

### DISCUSSION THREAD

**Discussion by vedantroy** [fastify/fastify#4994] [Q&A]
I have the following middleware:
```
const authPlugin = async (fastify: FastifyInstance, options: any) => {
  const authorize = async (req: FastifyRequest, reply: FastifyReply) => {
    const { userId } = checkIfUserIsAuthorized(req)

    if (!userId) {
      throw new Error("unauthorized")
    }

    let user = someDatabaseCall()
    req.user = user!
  }

  fastify.decorateRequest("user", null)
  fastify.decorate("authorize", authorize)
}
```
which checks if a user is authorized, and if so, attaches it to the requester.
The problem is, throwing unauthorized causes fastify to return a 500 error that is logged to the console. How can I tell fastify to not log this error? I would just return a 401, but I think this would allow the request to continue.

**Comment by metcoder95** [comment]
Based on my experience, I would recommend to not fork the flow of the request by allowing the `decorator` function to throw if the request is not following the business logic rules.

My recommendation here will be to do:
```ts
const authPlugin = async (fastify: FastifyInstance, options: any) => {
  const authorize = async (req: FastifyRequest, reply: FastifyReply) => {
    const { userId } = checkIfUserIsAuthorized(req)

    if (!userId) {
      return null;
    }

    // I might also move this to the route handler ( or preHandler if you prefer) instead
    let user = someDatabaseCall()
    req.user = user!
  }

  fastify.decorateRequest("user", null)
  fastify.decorate("authorize", authorize)
}
```

So, on the main route handle, you can apply the business rules accordingly, and decide how to respond to each missing piece.

However, if you still want to do it in that way, you can make use of [`fastify-sensible`](https://github.com/fastify/fastify-sensible) to throw the error that `fastify` can make use of to create a proper response.

**Reply by vedantroy** [reply]
Thanks, this worked well.

One problem is, we have a generic error handler:
```
app.setErrorHandler(async (error, request, reply) => {
  console.log(error)
  Sentry.captureException(error)
  reply.status(500).send({ error: "Something went wrong" })
})
```
which is capturing the error created by fastify-sensible, and logging it / returning with a 500 code.

How can we check if the error is a 400 error (e.g, thrown by fastify-sensible / expected), and if so, don't log it / report it to Sentry, etc.

**Reply by metcoder95** [reply]
A common way to do so is to check if the error contains `statusCode` property. e.g.
```js
app.setErrorHandler(async (error, request, reply) => {
  console.log(error)

// Just being careful of cleaning up the error to not leak sensitive information
  if (error.statusCode < 500) reply.status(error.statusCode).send(error);

  Sentry.captureException(error)
  reply.status(500).send({ error: "Something went wrong" })
})
```
